include/circt/Dialect/FIRRTL/FIRRTLExpressions.td

```diff
@@ -81,7 +87,7 @@ def SubindexOp : FIRRTLOp<"subindex", [NoSideEffect]> {
 
   // TODO: Could drop the result type, inferring it from the source.
   let assemblyFormat =
-     "$input `[` $index `]` attr-dict `:` functional-type($input, $result)";
+     "$input `[` $index `]` attr-dict `:` type($input)";
```

Comment:
One thing I ran into is that type constraints assume the input type has already been validated.  In these cases, you need to change:

` let arguments = (ins FIRRTLType:$input, UIntType:$index);`

to constrain `$input` to be an FIRRTLArrayType, otherwise `firrtl::getArrayElementType` in `IndexConstraint ` will fail and bad things will happen.